File: android/core/network/src/main/kotlin/xyz/klinik/network/AuthApi.kt
package xyz.klinik.network

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class LoginRequest(
    val identifier: String,
    val password: String,
    val totpCode: String? = null,
    val deviceName: String? = null,
    val platform: String? = "android",
)

@Serializable
data class LoginResponse(
    val status: Status,
    val accessToken: String? = null,
    val refreshToken: String? = null,
    val expiresIn: Int? = null,
    /**
     * Present only with MFA_SETUP_REQUIRED. Accepted solely by the enrolment
     * endpoints, and expires in five minutes.
     */
    val setupToken: String? = null,
) {
    @Serializable
    enum class Status {
        @SerialName("OK") OK,
        @SerialName("MFA_REQUIRED") MFA_REQUIRED,
        @SerialName("MFA_SETUP_REQUIRED") MFA_SETUP_REQUIRED,
    }

    /** Tokens, when the response actually carries a session. */
    fun tokens(nowMillis: Long = System.currentTimeMillis()): SessionTokens? {
        val access = accessToken ?: return null
        val refresh = refreshToken ?: return null
        val expires = expiresIn ?: return null

        return SessionTokens(access, refresh, nowMillis + expires * 1000L)
    }
}

@Serializable
data class TokensResponse(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Int,
) {
    fun tokens(nowMillis: Long = System.currentTimeMillis()): SessionTokens =
        SessionTokens(accessToken, refreshToken, nowMillis + expiresIn * 1000L)
}

@Serializable
data class TotpSetup(val secret: String, val uri: String)

/**
 * A device holding a session.
 *
 * `familyId` is what gets revoked: the server rotates refresh tokens and keeps
 * a family per sign-in, so ending one here ends that device rather than that
 * one token.
 */
@Serializable
data class SessionSummary(
    val familyId: String,
    val deviceName: String? = null,
    val platform: String? = null,
    val ipAddress: String? = null,
    val lastSeenAt: String,
    /** True for the device making this request. */
    val current: Boolean = false,
)

/**
 * An invitation, returned once.
 *
 * The code is shown exactly once because only its hash is stored — a screen
 * that expects to fetch it again later will show somebody a blank instead.
 */
@Serializable
data class Invitation(
    val id: String,
    val code: String,
    val expiresAt: String,
)

@Serializable
private data class ChangePasswordBody(val currentPassword: String, val newPassword: String)

@Serializable
private data class AcceptInvitationBody(
    val identifier: String,
    val code: String,
    val password: String,
)

@Serializable
private data class CreateInvitationBody(
    val email: String? = null,
    val phone: String? = null,
    val role: UserRole,
    val patientId: String? = null,
)

@Serializable
private data class TotpCode(val code: String)

/**
 * Authentication calls.
 *
 * Refresh lives in SessionManager rather than here: the backend treats refresh
 * tokens as single-use, so exactly one refresh may be in flight at a time and
 * that has to be enforced in one place.
 */
class AuthApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    suspend fun login(request: LoginRequest): LoginResponse {
        val body = client.send(
            Endpoint(
                method = HttpMethod.POST,
                path = "auth/login",
                body = json.encodeToString(request),
                requiresAuthentication = false,
            ),
        )

        return decode(body)
    }

    /**
     * Starts enrolment. Reached with the scoped setup token when the account
     * has no second factor yet, or with a session token when a patient opts in.
     */
    suspend fun beginTotpEnrolment(setupToken: String? = null): TotpSetup =
        decode(
            client.send(
                Endpoint(method = HttpMethod.POST, path = "auth/2fa/setup", bearerOverride = setupToken),
            ),
        )

    suspend fun confirmTotpEnrolment(code: String, setupToken: String? = null) {
        client.send(
            Endpoint(
                method = HttpMethod.POST,
                path = "auth/2fa/confirm",
                body = json.encodeToString(TotpCode(code)),
                bearerOverride = setupToken,
            ),
        )
    }

    /**
     * Turns two-factor off, with a current code.
     *
     * The code is required by the server and asked for here: somebody who has
     * walked away from an unlocked phone should not be able to remove the
     * second factor from the account they left open.
     */
    suspend fun disableTotp(code: String) {
        client.send(
            Endpoint(
                method = HttpMethod.POST,
                path = "auth/2fa/disable",
                body = json.encodeToString(TotpCode(code)),
            ),
        )
    }

    /** Changing a password. The current one is required; the server checks it. */
    suspend fun changePassword(currentPassword: String, newPassword: String) {
        client.send(
            Endpoint(
                method = HttpMethod.POST,
                path = "auth/password",
                body = json.encodeToString(
                    ChangePasswordBody.serializer(),
                    ChangePasswordBody(currentPassword, newPassword),
                ),
            ),
        )
    }

    /** Every device holding a session, with this one marked. */
    suspend fun sessions(): List<SessionSummary> =
        decode(client.send(Endpoint(method = HttpMethod.GET, path = "auth/sessions")))

    /** Ends one device's session. */
    suspend fun endSession(familyId: String) {
        client.send(Endpoint(method = HttpMethod.DELETE, path = "auth/sessions/$familyId"))
    }

    /** Ends every session, including this one. */
    suspend fun signOutEverywhere() {
        client.send(Endpoint(method = HttpMethod.POST, path = "auth/logout-all"))
    }

    /**
     * Invites somebody. The code comes back once and is delivered by the
     * clinic — the server keeps only its hash.
     */
    suspend fun invite(
        role: UserRole,
        email: String? = null,
        phone: String? = null,
        patientId: String? = null,
    ): Invitation =
        decode(
            client.send(
                Endpoint(
                    method = HttpMethod.POST,
                    path = "auth/invitations",
                    body = json.encodeToString(
                        CreateInvitationBody.serializer(),
                        CreateInvitationBody(email, phone, role, patientId),
                    ),
                ),
            ),
        )

    /** Redeems an invitation and signs the new account in. */
    suspend fun acceptInvitation(
        identifier: String,
        code: String,
        password: String,
    ): LoginResponse =
        decode(
            client.send(
                Endpoint(
                    method = HttpMethod.POST,
                    path = "auth/invitations/accept",
                    body = json.encodeToString(
                        AcceptInvitationBody.serializer(),
                        AcceptInvitationBody(identifier, code, password),
                    ),
                ),
            ),
        )

    suspend fun signOut() {
        client.send(Endpoint(method = HttpMethod.POST, path = "auth/logout"))
    }

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}

File: ios/Sources/KlinikAPI/AuthAPI.swift
import Foundation
import KlinikCore

public struct LoginRequest: Encodable, Sendable {
    public let identifier: String
    public let password: String
    public let totpCode: String?
    public let deviceName: String?
    public let platform: String?

    public init(
        identifier: String,
        password: String,
        totpCode: String? = nil,
        deviceName: String? = nil,
        platform: String? = "ios"
    ) {
        self.identifier = identifier
        self.password = password
        self.totpCode = totpCode
        self.deviceName = deviceName
        self.platform = platform
    }
}

public struct LoginResponse: Decodable, Sendable {
    public enum Status: String, Decodable, Sendable {
        case ok = "OK"
        case mfaRequired = "MFA_REQUIRED"
        case mfaSetupRequired = "MFA_SETUP_REQUIRED"
    }

    public let status: Status
    public let accessToken: String?
    public let refreshToken: String?
    public let expiresIn: Int?
    /// Present only with `mfaSetupRequired`. Accepted solely by the enrolment
    /// endpoints, and expires in five minutes.
    public let setupToken: String?

    /// Tokens, when the response actually carries a session.
    public func tokens(now: Date = Date()) -> SessionTokens? {
        guard let accessToken, let refreshToken, let expiresIn else {
            return nil
        }

        return SessionTokens(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAt: now.addingTimeInterval(TimeInterval(expiresIn))
        )
    }
}

public struct TokensResponse: Decodable, Sendable {
    public let accessToken: String
    public let refreshToken: String
    public let expiresIn: Int

    public func tokens(now: Date = Date()) -> SessionTokens {
        SessionTokens(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAt: now.addingTimeInterval(TimeInterval(expiresIn))
        )
    }
}

public struct DeviceSession: Decodable, Sendable, Equatable, Identifiable {
    public let familyId: String
    public let deviceName: String?
    public let platform: String?
    public let ipAddress: String?
    public let lastSeenAt: Date
    public let current: Bool

    public var id: String { familyId }
}

public struct TotpSetup: Decodable, Sendable {
    public let secret: String
    public let uri: String

    /// The secret in groups of four, which is how somebody types it into an
    /// authenticator by hand without losing their place.
    public var groupedSecret: String {
        stride(from: 0, to: secret.count, by: 4)
            .map { offset -> String in
                let start = secret.index(secret.startIndex, offsetBy: offset)
                let end = secret.index(start, offsetBy: min(4, secret.count - offset))

                return String(secret[start..<end])
            }
            .joined(separator: " ")
    }
}

/// An invitation, returned once (spec M1).
///
/// The code is shown a single time — only its hash is kept — so the screen
/// that receives it has to be the screen that delivers it.
public struct Invitation: Decodable, Sendable, Equatable {
    public let id: String
    public let code: String
    public let expiresAt: Date
}

/// Authentication calls.
///
/// Refresh lives here too, but it is reached through `SessionManager` rather
/// than called directly: the backend treats refresh tokens as single-use, so
/// exactly one refresh may be in flight at a time.
public struct AuthAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func login(_ request: LoginRequest) async throws -> LoginResponse {
        try await client.send(
            Endpoint(
                method: .post,
                path: "auth/login",
                body: try JSONEncoder.klinik.encode(request),
                requiresAuthentication: false
            ),
            as: LoginResponse.self
        )
    }

    /// Starts enrolment. Reached with the scoped setup token when the account
    /// has no second factor yet, or with a session token when a patient opts in.
    public func beginTotpEnrolment(setupToken: String? = nil) async throws -> TotpSetup {
        try await client.send(
            Endpoint(method: .post, path: "auth/2fa/setup", bearerOverride: setupToken),
            as: TotpSetup.self
        )
    }

    public func confirmTotpEnrolment(code: String, setupToken: String? = nil) async throws {
        try await client.send(
            Endpoint(
                method: .post,
                path: "auth/2fa/confirm",
                body: try JSONEncoder.klinik.encode(["code": code]),
                bearerOverride: setupToken
            )
        )
    }

    public func sessions() async throws -> [DeviceSession] {
        try await client.send(Endpoint(method: .get, path: "auth/sessions"), as: [DeviceSession].self)
    }

    public func signOut() async throws {
        try await client.send(Endpoint(method: .post, path: "auth/logout"))
    }

    public func signOutEverywhere() async throws {
        try await client.send(Endpoint(method: .post, path: "auth/logout-all"))
    }

    /**
     * Invites somebody to open an account (spec M1).
     *
     * `patientId` links the new login to a file that already exists, which is
     * the normal case: the clinic opens the file when the patient books, and
     * the account comes later.
     */
    public func invite(
        email: String?,
        phone: String?,
        role: String,
        patientId: String?
    ) async throws -> Invitation {
        try await client.send(
            Endpoint(
                method: .post,
                path: "auth/invitations",
                body: try JSONEncoder.klinik.encode(
                    InviteBody(email: email, phone: phone, role: role, patientId: patientId)
                )
            ),
            as: Invitation.self
        )
    }

    public func revokeSession(familyId: String) async throws {
        try await client.send(Endpoint(method: .delete, path: "auth/sessions/\(familyId)"))
    }
}

/// Refreshes without going through `APIClient`, which would need a valid token
/// to obtain one — a loop.
private struct InviteBody: Encodable {
    let email: String?
    let phone: String?
    let role: String
    let patientId: String?
}

public struct HTTPTokenRefresher: TokenRefresher {
    private let baseURL: URL
    private let transport: HTTPTransport

    public init(baseURL: URL, transport: HTTPTransport) {
        self.baseURL = baseURL
        self.transport = transport
    }

    public func refresh(using refreshToken: String) async throws -> SessionTokens {
        var request = URLRequest(url: baseURL.appendingPathComponent("auth/refresh"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder.klinik.encode(["refreshToken": refreshToken])

        let response = try await transport.send(request)

        guard (200...299).contains(response.status) else {
            let body = (try? JSONDecoder.klinik.decode(ErrorResponse.self, from: response.body))
                ?? ErrorResponse(statusCode: response.status, message: "")
            throw APIError.from(status: response.status, body: body)
        }

        return try JSONDecoder.klinik.decode(TokensResponse.self, from: response.body).tokens()
    }
}
